tests/integ/compiler/test_query_generator.py

```diff
@@ -0,0 +1,226 @@
+    # unpivot
+    df_unpivot = session.create_dataframe(
+        [(1, "electronics", 100, 200), (2, "clothes", 100, 300)],
+        schema=["empid", "dept", "jan", "feb"],
+    ).unpivot("sales", "month", ["jan", "feb"])
+    check_generated_plan_queries(df_unpivot._plan)
+
+
+def test_groupby_agg(session):
+    df = session.create_dataframe([[1, 2], [3, 4]], schema=["a", "b"])
+    df_groupby = df.group_by("a").avg("b").filter(col("a") == 1)
+    check_generated_plan_queries(df_groupby._plan)
+
+
+def test_window_function(session):
+    window1 = (
+        Window.partition_by("value").order_by("key").rows_between(Window.CURRENT_ROW, 2)
+    )
+    window2 = Window.order_by(col("key").desc()).range_between(
+        Window.UNBOUNDED_PRECEDING, Window.UNBOUNDED_FOLLOWING
+    )
+    df = (
+        session.create_dataframe(
+            [(1, "1"), (2, "2"), (1, "3"), (2, "4")], schema=["key", "value"]
+        )
+        .select(
+            avg("value").over(window1).as_("window1"),
+            avg("value").over(window2).as_("window2"),
+        )
+        .sort("window1")
+    )
+    df_result = df.union_all(df).select("*")
+    check_generated_plan_queries(df_result._plan)
+
+
+def test_multiple_plan_query_generation(session):
+    df = session.create_dataframe([[1, 2], [3, 4]], schema=["a", "b"])
+    table_name = random_name_for_temp_object(TempObjectType.TABLE)
+    with session.query_history() as query_history:
+        df.write.save_as_table(
+            table_name,
+            table_type="temp",
+            mode="overwrite",
+        )
+    # get the last query executed
+    table_create_query = query_history.queries[-1].sql_text
+    df2 = df.select((col("a") + 1).as_("a"), "b")
+    df_res = df.union_all(df2)
+    df_queries = [query.sql for query in df_res._plan.queries]
+    df_post_actions = [query.sql for query in df_res._plan.post_actions]
+    expected_queries = [table_create_query] + df_queries
+    expected_post_actions = df_post_actions
+
+    create_table_logic_plan = SnowflakeCreateTable(
+        [table_name],
+        column_names=None,
+        mode=SaveMode.OVERWRITE,
+        query=df._plan,
+        table_type="temp",
+        clustering_exprs=None,
+        comment=None,
+    )
+    snowflake_plan = session._analyzer.resolve(create_table_logic_plan)
+    query_generator = create_query_generator(snowflake_plan)
+    reset_plan_tree(snowflake_plan)
+    reset_plan_tree(df_res._plan)
+    logical_plans = [snowflake_plan.source_plan, df_res._plan.source_plan]
+    generated_queries = query_generator.generate_queries(logical_plans)
+    result_queries = [
+        query.sql.lstrip() for query in generated_queries[PlanQueryType.QUERIES]
+    ]
+    result_post_actions = [
+        query.sql.lstrip() for query in generated_queries[PlanQueryType.POST_ACTIONS]
+    ]
+    assert result_queries == expected_queries
+    assert result_post_actions == expected_post_actions
```

Comment:
can we add a test for 1) non-select query with session.sql 2) another type of dataframe operation with multiple queries, like https://github.com/snowflakedb/snowpark-python/blob/50275a7e0c93035759fd95f2563f18d99ae46954/tests/integ/test_dataframe.py#L2995 3) dataframe reader https://github.com/snowflakedb/snowpark-python/blob/50275a7e0c93035759fd95f2563f18d99ae46954/tests/integ/test_cte.py#L389-L390

Reply:
added